### PR TITLE
Support annotations for versioning behavior

### PR DESCRIPTION
**Pull request by antlai-temporal**
## What was changed

This PR supports providing a default versioning behavior in Worker Options, and a programmatic API to override these defaults:

For example:
``` go
	workflow.SetVersioningBehavior(ctx, workflow.VersioningBehaviorPinned)
```
called  during the first workflow task sets the versioning mode for that workflow to `Pinned`, eliminating the need of patching.

And 
```go
	w := worker.New(c, ts.taskQueueName, worker.Options{
		DeploymentOptions: worker.DeploymentOptions{
                            BuildID:                   "1.0",
		            UseBuildIDForVersioning:   true,
		            DeploymentName:            "deploy-test1",
		            DefaultVersioningBehavior: workflow.VersioningBehaviorAutoUpgrade,
              }
	})
```
sets the default versioning behavior to `AutoUpgrade`. Note that `DeploymentName`, `UseBuildIDForVersioning` and `BuildID` need to be set to enable versioning-3.
If versioning-3 is on, but no default or programmatic versioning behavior has been set for a workflow, the server will fail its first workflow task. 

## Why?

A hint that a workflow is short-running, and therefore, does not need to be moved to the new Build ID on the next task, eliminates the need of patching for many workflow types. 

## Checklist

1. Closes 
#1663 
2. How was this tested:
Integration test that uses versioning-2 and a grpc interceptor to show that versioning behaviors are propagated with responses.
The dependency on versioning-2 is just to set a default Build ID.

### CODE REVIEW

File: workflow/workflow.go
Comment (cretz) [inline workflow/workflow.go:638]: I do not believe this is needed. Just use the worker option. I don't think we need global default for this worker option any more than any other (if a user wants to reference a global `EnableLoggingInReplay` or `VersioningBehavior` or whatever when they make their workers they can). This will because especially true as you get to other SDKs where we don't like customizable global defaults.

File: internal/workflow.go
Comment (cretz) [inline internal/workflow.go:69]: Hrmm, I wonder if we want to expose the raw proto enum like we do with `StartWorkflowOptions.WorkflowIDConflictPolicy` or not like I guess we chose not with `WorkflowUpdateStage`?

File: test/integration_test.go
Comment (cretz) [inline test/integration_test.go:6219]: This seems to be the only use of the "from proto" exposed utility, but there doesn't seem to be much benefit converting in this test. Can you just use the raw proto enum in this test for this slice and not have the utility?

File: internal/workflow.go
Comment (cretz) [inline internal/workflow.go:2558]:
```suggestion
func versioningBehaviorToProto(t VersioningBehavior) enumspb.VersioningBehavior {
```

File: internal/internal_workflow.go
Comment (Quinn-With-Two-Ns) [inline internal/internal_workflow.go:1614]: Did we intentionally omit a getter here?

File: workflow/workflow.go
Comment (Sushisource) [reply workflow/workflow.go:638]: Also agree

Comment (antlai-temporal) [reply workflow/workflow.go:638]: This is not global, it only changes the running workflow options, so that we can override  the default for a particular workflow. I agree about no process-level defaults...

File: internal/workflow.go
Comment (antlai-temporal) [reply internal/workflow.go:69]: We could, what do we want to do moving forward @Quinn-With-Two-Ns , wrap enums or raw protos?

File: test/integration_test.go
Comment (antlai-temporal) [reply test/integration_test.go:6219]: I think the bigger question is your previous comment, whether we just expose the enum protos and be done with it. If not I think sooner or later users may want to go back and forth, so more boiler plate...

File: internal/internal_workflow.go
Comment (antlai-temporal) [reply internal/internal_workflow.go:1614]: Yes, versioning behavior  can be dependent on worker options so if app code branches based on it they will get an NDE. And in general, we wanted to be able to change annotations without needing patching, so no-getter helps there...

File: internal/workflow.go
Comment (Quinn-With-Two-Ns) [reply internal/workflow.go:69]: Yeah there is pros and cons to both and the SDK has historically not been consistent. I would bias towards wrapping the enums.

File: internal/internal_workflow.go
Comment (Quinn-With-Two-Ns) [reply internal/internal_workflow.go:1614]: Is the versioning behaviour written into history?

File: internal/internal_worker.go
Comment (ShahabT) [inline internal/internal_worker.go:1674]: As also discussed offline this morning, I think it's better to wrap these options in a struct called `DeploymentOption` or something. The list may grow in the future and we don't want the user to have to repeat all the fields individually in each TQ. Example:
```
deploymentOptions := worker.DeploymentOptions{
	DeploymentName: "my-app",
	BuildID: os.Getenv("BUILD_ID"),
	UseBuildIDForVersioning: true,
	DefaultVersioningBehavior: workflow.VersioningBehaviorAutoUpgrade,
}

w1 := worker.NewWorker(client, "foo", worker.WorkerOptions{
	DeploymentOptions: deploymentOptions,
})

w2 := worker.NewWorker(client, "bar", worker.WorkerOptions{
	DeploymentOptions: deploymentOptions,
})
```

File: workflow/workflow.go
Comment (cretz) [reply workflow/workflow.go:638]: I see, my mistake. I see this accepts a workflow context. I misread, sorry.

File: internal/workflow.go
Comment (cretz) [reply internal/workflow.go:69]: Works for me

File: internal/internal_workflow.go
Comment (cretz) [reply internal/internal_workflow.go:1614]: I think you need to be able to get something you may have set. I think this should either be in workflow register options and/or this should be called `SetVersioningBehaviorOverride` (and `GetVersioningBehaviorOverride` will just return what you have called with the "set", not the resolved value).

File: internal/internal_worker.go
Comment (antlai-temporal) [reply internal/internal_worker.go:1674]: I'm happy with this. @cretz and @Quinn-With-Two-Ns shall I add the struct and deprecate the original BuildID and UseBuildIdForVersioning?

File: internal/internal_workflow.go
Comment (antlai-temporal) [reply internal/internal_workflow.go:1614]: @Quinn-With-Two-Ns My understanding is that it is not part of the history, at least I do nothing special during replay, @ShahabT can you confirm?

Comment (antlai-temporal) [reply internal/internal_workflow.go:1614]: @cretz we could add the Override suffix, but the emphasis was that most customers will annotate all workflows, and do not set defaults in Workers, since that is safer.
 The Get override, could be useful in the future when we enable changing workflow properties externally (like a reset), maybe we could add it then.
I think the problem of adding to workflow register is that I don't have the context of the running workflow to latch on, so I cannot change the behavior based on arguments, or for dynamic workflows. Also, it may be more clear to have this "annotation" the first line of the workflow definition, together with the code... 
@drewhoskins-temporal any preferences?

File: internal/internal_worker.go
Comment (Quinn-With-Two-Ns) [reply internal/internal_worker.go:1674]: All these fields are marked experimental right? If so I have no strong opinion here. If we do this in Go we should probably group them in all SDKs though.

File: test/integration_test.go
Comment (antlai-temporal) [reply test/integration_test.go:6219]: Ok, I don't think implementing the FromProto if they need it is a big deal. I'm getting rid of it, using proto in the test, and making the toProto private.

Comment (cretz) [reply test/integration_test.go:6219]: Works for me. Yeah the high-level wrapper is nicer for users, and no need to convert the other way.

File: internal/internal_workflow.go
Comment (antlai-temporal) [reply internal/internal_workflow.go:1614]: Looking at the proto changes that @ShahabT did, the versioning behavior has been added to WorkflowTaskCompletedEventAttributes in history. I don't do anything with it (maybe I should?) but that would give the server a placeholder to record the versioning behavior that I send after each task...

Comment (cretz) [reply internal/internal_workflow.go:1614]: I do wonder whether the getter should take that into account. I'll defer to y'all on desired behavior of the getter, but I do believe we should not have a setter without a getter (even if we have to add caveats to the behavior of the getter).

Comment (carlydf) [reply internal/internal_workflow.go:1614]: The server stores the current build id and versioning behavior in Mutable State and in visibility. The previously-used build ids are also stored in visibility -- I'm not sure if previously-used build ids are also stores in Mutable State.

"getting" the current versioning behavior of a given workflow execution could be done by reading the mutable state or visibility entry.

Querying a bunch of workflow executions for a specific versioning behavior and/or build id combination would be a visibility query (we will use this for reachability).

It seems to me that to support a workflow-execution-level versioning behavior override, the server would need to store not just the (versioning behavior, build id) tuple, but also an `is_override` (or some other name) bit, that basically means, `if is_override, ignore whatever the sdk says about my workflow type and do task routing based on what is currently written in my mutable state as an override`. To me it seems like it could be handled on the server side, since that's where the task routing decision is made.

**Comment by antlai-temporal** [comment]
I think the latest commit addresses most of the issues raised, let me know if it is ready to merge...

File: internal/internal_workflow.go
Comment (cretz) [reply internal/internal_workflow.go:1614]: > I think of versioning behavior as primarily a workflow type level thing, with a workflow worker level default in case it's not specified at the type level (but from a product perspective, we want to encourage explicitly setting it for each workflow type

If this is the case, it also needs to be added to `workflow.RegisterOptions`.

> "getting" the current versioning behavior of a given workflow execution could be done by reading the mutable state or visibility entry.

We don't want to get from server what's not already in history. We just need to make sure any `SetWhatever` _inside_ a workflow has a corresponding `GetWhatever`. If this just returns the value that was "set" before in the same workflow code and not the actual server value, that's fine (though it does seem like task complete may have some form of the value). The docs of the getter just need to clarify.

**Comment by cretz** [comment]
@antlai-temporal - based on thread above, I wonder if this needs to be on `worker.RegisterOptions` too if that is the most common way this should be set (at the workflow type level).

**Comment by antlai-temporal** [comment]
> @antlai-temporal - based on thread above, I wonder if this needs to be on `worker.RegisterOptions` too if that is the most common way this should be set (at the workflow type level).

We could have a type-level default in RegisterOptions that can be overridden by the programmatic API. That would be more consistent with the Java annotation+programmatic override I was thinking on. The Getter will probably return just the programmatic to be consistent that we can change worker defaults without patching.
I'm trying to unblock testing for the server team, so I may do that in a separate PR if that's ok...

File: workflow/workflow.go
Comment (Quinn-With-Two-Ns) [inline workflow/workflow.go:629]: What if it is called, but not in the first workflow task?

Comment (antlai-temporal) [reply workflow/workflow.go:629]: If there was no worker level default, the server will fail the first workflow task, so there is no progress until a worker sets a default.
If tthere was a worker level default, it will change it ok.
In the comment I'm encouraging to set it in the first task to avoid the first case...

**Comment by cretz** [comment]
> I'm trying to unblock testing for the server team, so I may do that in a separate PR if that's ok...

I think it's probably good (will defer to @Quinn-With-Two-Ns) but probably want to create an issue to track the work. I do think we should be encouraging that way to set worker-type level value and this should be a lesser-used in-workflow override.

>  In the comment I'm encouraging to set it in the first task to avoid the first case...

This is why we should be encouraging the workflow register options approach over the inline approach (can even change the name of the inline one to something less wieldy, e.g. `SetVersioningBehaviorOverride`).

**Comment by Quinn-With-Two-Ns** [comment]
Yeah OK merging into the branch, +1 to everything Chad said.

Once we are further along I would also like to see documentation as to how this workflow API is supposed to interact with versioning info set externally described here https://github.com/temporalio/sdk-go/pull/1692#discussion_r1821431989.

**Review by Quinn-With-Two-Ns — APPROVED** [review]

**Comment by antlai-temporal** [comment]
> Yeah OK merging into the branch, +1 to everything Chad said.
> 
> Once we are further along I would also like to see documentation as to how this workflow API is supposed to interact with versioning info set externally described here [#1692 (comment)](https://github.com/temporalio/sdk-go/pull/1692#discussion_r1821431989).

@cretz @Quinn-With-Two-Ns 
I added the issue https://github.com/temporalio/sdk-go/issues/1698 . Merging as it is to unblock server team, but will add another PR before merging to main.
We will clarify the interactions with the external workflow property setting (after discussing today with @carlydf they seem orthogonal). The keyword "override" is used by this external mechanism, I will stick to default vs non-default to avoid confusion. The getter will only return what it was set (as it does right now).

**Pull request merged**
